lib/Parse/ParseExpr.cpp

```diff
@@ -3581,26 +3581,22 @@ Parser::parseExprCollectionElement(Optional<bool> &isDictionary) {
     return Element;
 
   // Parse the ':'.
-  if (!consumeIf(tok::colon)) {
-    if (Element.hasCodeCompletion()) {
-      // Return the completion expression itself so we can analyze the type
-      // later.
-      return Element;
+  ParserResult<Expr> Value;
+  if (consumeIf(tok::colon)) {
+    // Parse the value.
+    Value = parseExpr(diag::expected_value_in_dictionary_literal);
+    if (Value.isNull()) {
+      if (!Element.hasCodeCompletion()) {
+        Value = makeParserResult(Value, new (Context) ErrorExpr(PreviousLoc));
+      } else {
+        Value = makeParserResult(Value,
+                                 new (Context) CodeCompletionExpr(PreviousLoc));
+      }
     }
+  } else {
     diagnose(Tok, diag::expected_colon_in_dictionary_literal);
-    return ParserStatus(Element) | makeParserError();
-  }
-
-  // Parse the value.
-  auto Value = parseExpr(diag::expected_value_in_dictionary_literal);
-
-  if (Value.isNull()) {
-    if (!Element.hasCodeCompletion()) {
-      Value = makeParserResult(Value, new (Context) ErrorExpr(PreviousLoc));
-    } else {
-      Value = makeParserResult(Value,
-                               new (Context) CodeCompletionExpr(PreviousLoc));
-    }
+    Value = makeParserResult(makeParserError(),
+                             new (Context) ErrorExpr(SourceRange()));
```

Comment:
@rintaro is making an ErrorExpr with an invalid source range for a "missing" value ok? This was needed for test cases like `let x: [EnumA:String] = [.a: "foo", foo.<complete>]` so that constraint generation treats the last element as a key. I don't have any appropriate source location we can give it. The PreviousLoc is the key, so if we used that the value's source range ends up overlapping the key's which seems bad, and the loc after the key may be EOF, ], or something random. The other option is to just put the element in as is and update either constraint generation or solving (like we did for dictionary/array literal mismatch), but I'm not sure how many places assume the elements of DictionaryExpr are TupleExprs.

Reply:
I *think* we can use a invalid source range here. As far as I remember, in code completion, nothing uses the range of the `ErrorExpr`. Also `TupleExpr::getSourceRange()` correctly ignores invalid element source ranges so diagnostics shouldn't be affected.

Reply:
Do you think it would make sense to introduce a special `HoleExpr` as a placeholder for missing AST elements just like we introduce `HoleType` into the constraint system to indicate that something couldn't be inferred?